helix-term/src/ui/completion.rs

```diff
@@ -213,14 +227,27 @@ impl Completion {
 
             let (view, doc) = current!(editor);
 
-            // if more text was entered, remove it
-            doc.restore(view, &savepoint);
-
             match event {
-                PromptEvent::Abort => {
-                    editor.last_completion = None;
-                }
+                PromptEvent::Abort => {}
                 PromptEvent::Update => {
+                    // Update creates "ghost" transactiosn which are not send to the
+                    // lsp server to avoid messing up rerequesting completions. Once a
+                    // completion has been selected (with) tab it's always accepted whenever anything
+                    // is typed. The only way to avoid that is to explicitly abort the completion
+                    // with esc/c-c. This will remove the "ghost" transaction.
+                    //
+                    // The ghost transaction is modeled with a transaction that is not send to the LS.
+                    // (apply_temporary) and a savepoint. It's extremly important this savepoint is restored
+                    // (also without sending the transaction to the LS) *before any further transaction is applied*.
+                    // Otherwise incremental sync breaks (since the state of the LS doesn't match the state the transaction
+                    // is applied to).
+                    if editor.last_completion.is_none() {
+                        editor.last_completion = Some(CompleteAction::Selected {
+                            savepoint: doc.savepoint(view),
+                        })
+                    }
+                    // if more text was entered, remove it
+                    doc.restore(view, &savepoint, false);
```

Comment:
Is there mechanism to tell the language server to abort?

Reply:
as with most things completions are entirely managed by the client so no there isn't